### DISCUSSION TITLE
Scale a Milvus cluster on GCP

### DISCUSSION THREAD

**Discussion by buntys2010** [milvus-io/milvus#23855] [Q&A and General discussion] [answered]
I have a milvus cluster set-up on GCP using this article https://milvus.io/docs/gcp.md

I have loaded around 20 million embeddings into milvus and now it has stopped responding to the search queries or takes a long time to respond.

I tried to add another node and run below command (https://milvus.io/docs/scaleout.md)

helm upgrade history-documents-release milvus/milvus --set queryNode.replicas=3 --reuse-values

But it seems its not using the 3rd node at all, how can i solve this error?

**Comment by yhmo** [comment]
Some questions:
- vector dimension is?
- static data or dynamic data(continually insertion and deletion)?
- how many search requests per minute? the average latency of each search is?
- any configurations changed to the milvus service?

**Reply by buntys2010** [reply]
- Vector dimension is 768
- We don't delete any index but we are adding another 5 million embeddings (Continuous insert, but when i am trying to insert it doesnt connect at all or fails with error below)
- As of now we have almost zero to few search in a day
- Its a default configuration 

Error
```
RPC error: [get_loading_progress], <MilvusException: (code=1, message=collection 440065816806046537 has not been loaded to memory or load failed)>, <Time:{'RPC start': '2023-05-02 13:32:29.167890', 'RPC error': '2023-05-02 13:32:29.206872'}>
RPC error: [wait_for_loading_collection], <MilvusException: (code=1, message=collection 440065816806046537 has not been loaded to memory or load failed)>, <Time:{'RPC start': '2023-05-02 13:17:14.377496', 'RPC error': '2023-05-02 13:32:29.207147'}>
RPC error: [load_collection], <MilvusException: (code=1, message=collection 440065816806046537 has not been loaded to memory or load failed)>, <Time:{'RPC start': '2023-05-02 13:17:14.291600', 'RPC error': '2023-05-02 13:32:29.207266'}>
Traceback (most recent call last):
  File "load_milvus_history_script-2.py", line 133, in <module>
    mycollection.load()
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/orm/collection.py", line 361, in load
    conn.load_collection(self._name, replica_number=replica_number, timeout=timeout, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 109, in handler
    raise e
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 105, in handler
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 136, in handler
    ret = func(self, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 85, in handler
    raise e
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 50, in handler
    return func(self, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/client/grpc_handler.py", line 660, in load_collection
    self.wait_for_loading_collection(collection_name, timeout)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 109, in handler
    raise e
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 105, in handler
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 136, in handler
    ret = func(self, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 85, in handler
    raise e
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 50, in handler
    return func(self, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/client/grpc_handler.py", line 678, in wait_for_loading_collection
    progress = self.get_loading_progress(collection_name, timeout=timeout)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 109, in handler
    raise e
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 105, in handler
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 136, in handler
    ret = func(self, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 85, in handler
    raise e
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/decorators.py", line 50, in handler
    return func(self, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pymilvus/client/grpc_handler.py", line 739, in get_loading_progress
    raise MilvusException(response.status.error_code, response.status.reason)
pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=collection 440065816806046537 has not been loaded to memory or load failed)>
```

**Reply by yhmo** [reply]
Did you call collection.load()?

**Reply by buntys2010** [reply]
Here is the code snippet i am using to load index (At this point it gets stuck in loading the collection - i tried and waited for 30 mins)

First i am loading the collection

```
if utility.has_collection(f"{collection_name}"):
    print(f"Connecting to collection '{collection_name}'")
    collection = Collection(collection_name)
    collection.load()
else:
    print(f"Creating collection '{collection_name}'")
```

Once the collection is loaded, using below code to load index

```
mr = collection.insert(data_to_insert)
ids = mr.primary_keys
collection.release()

print(fmt.format("Start Creating index IVF_SQ8"))
index_param = {
    "index_type": "IVF_SQ8",
    "metric_type": "IP",
    "params": {"nlist": 4096},
}

collection.create_index(field_name="embedding", index_params=index_param)

print("Sleeping for few secs")
time.sleep(50)
```

**Reply by yhmo** [reply]
20M rows, dimension=768, the original data size is 768*4*20M=60GB. 

If you have called collection.load(), but it still returns the error "collection 440065816806046537 has not been loaded to memory or load failed", I guess the memory capacity of query nodes is insufficient.

**Reply by yhmo** [reply]
How much RAM is assigned to each query node?

**Reply by buntys2010** [reply]
Yes i understand that there might be a memory load on the server.

1. Currently I am unable to load the collection and its stuck, i am not receiving any errors (even after 5 mins)

3. I am currently using **e2-standard-8** machines with has 32 GB RAM.
4. Initially i was using 2 nodes and i recently introduced 3rd node.
5. I added the node using this blog https://milvus.io/docs/scaleout.md but it seems the 3rd node is not being used as per below screen shot.

![image](https://user-images.githubusercontent.com/9404463/236161692-30b83572-fa37-49e8-8ec9-efb40aab5b08.png)

**Reply by yhmo** [reply]
Could you try below steps:
1. release the collection 'AR_2021'
2. load the collection again

**Reply by yhmo** [reply]
Your milvus version is?

**Reply by buntys2010** [reply]
Apologies for delayed response, I am using version 2.2.8.

Can you guide me how to release a collection without loading via a python or CLI?

**Reply by buntys2010** [reply]
@yhmo Can you advise me how to release the collection without loading? Currently i cant load the collection even after 30 mins. It gets stuck on this line and unable to move further from this.

`collection = Collection(collection_name)`

**Reply by yhmo** [reply]
Being stuck at this line is weird, the `Collection(collection_name)` firstly calls `has_collection()` to check the collection existence then calls `describe_collection()` to get the collection schema. The two apis are light and should not be blocked.
Restart the milvus service, and observe the memory usage of the 3 query nodes. Ideally, after restarting, the milvus will automatically load the collection into the 3 query nodes.

**Reply by buntys2010** [reply]
@yhmo Do you have documentation on how to restart milvus service without using loosing existing set-up, my apologies i am new to Milvus. Also its bit unexpected behavior as this is happening just for one particular collection and i am able to access other collection without any issue.

**Reply by LoveEachDay** [reply, marked as answer]
You can restart your milvus service via:
```
kubectl get deploy | grep history-documents-release-milvus | awk '{print $1}' | xargs kubectl rollout restart deploy 
```

**Reply by buntys2010** [reply]
thank you @LoveEachDay and @yhmo for your help, i am able to access the collection successfully now.

**Comment by divyajaincs** [comment]
@LoveEachDay can we enable the autoscaling on milvus cluster(without manually setting up the replicas) which is deployed on GCP
